File: Class 19/dmopc17c1p3.java
import java.io.*;
import java.util.*;

public class dmopc17c1p3{
  public static void main(String[] args) throws IOException{
    BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
    
    String[] tokens = in.readLine().split(" ");
    int N = Integer.parseInt(tokens[0]), M = Integer.parseInt(tokens[1]);
    boolean[] vist = new boolean[N + 1];
    ArrayList<NodeAndDist>[] adj = new ArrayList[N + 1];
    for(int i = 1; i <= N; i++) adj[i] = new ArrayList<NodeAndDist>();
    final int INFINITY = 1000 * 1000;
    
    for(int i = 0; i < M; i++){
      tokens = in.readLine().split(" ");
      int node1 = Integer.parseInt(tokens[0]), node2 = Integer.parseInt(tokens[1]), dang = Integer.parseInt(tokens[2]);
      
      adj[node1].add(new NodeAndDist(node2, new Distance(dang, 1)));
      adj[node2].add(new NodeAndDist(node1, new Distance(dang, 1)));
    }
    
    PriorityQueue<NodeAndDist> heap = new PriorityQueue<NodeAndDist>();
    Distance[] dist = new Distance[N + 1];
    for(int i = 2; i <= N; i++)
      dist[i] = new Distance(INFINITY, INFINITY);
    dist[1] = new Distance(0, 0);
    heap.offer(new NodeAndDist(1, dist[1]));
    
    while(!heap.isEmpty()){
      NodeAndDist lowest = heap.poll();
      int current = lowest.node;
      if(vist[current]) continue;
      vist[current] = true;
      
      for(NodeAndDist to : adj[current]){
        int neighbour = to.node;
        Distance weight = to.dist;
        
        Distance newDist 
          = new Distance(dist[current].dang + weight.dang, dist[current].total + weight.total);
        
        if(newDist.compareTo(dist[neighbour]) < 0){
          dist[neighbour] = newDist;
          heap.offer(new NodeAndDist(neighbour, newDist));
        }
      }
    }
    
    if(dist[N].total == INFINITY)
      System.out.println(-1);
    else
      System.out.println(dist[N].dang + " " + dist[N].total);
  }
}

class Distance implements Comparable<Distance>{
  Integer dang, total;
  
  public Distance(int d, int t){
    this.dang = d;
    this.total = t;
  }
  
  public int compareTo(Distance other){
    if(this.dang == other.dang)
      return this.total.compareTo(other.total);
    return this.dang.compareTo(other.dang);
  }
}

class NodeAndDist implements Comparable<NodeAndDist>{
  int node;
  Distance dist;
  
  public NodeAndDist(int n, Distance d){
    this.node = n;
    this.dist = d;
  }
  
  public int compareTo(NodeAndDist other){
    return this.dist.compareTo(other.dist);
  }
}

File: Class 19/dmopc17c1p3.cpp
#include <stdio.h>
#include <vector>
#include <queue>
using namespace std;
const int MN = 1e5+5, INF = 1e6;

struct Distance{
    int dang, total;
};

struct NodeAndDist{
    int node;
    Distance dist;
};

vector<NodeAndDist> adj[MN];
Distance dist[MN];

bool operator < (Distance a, Distance b){
    if(a.dang == b.dang) return a.total < b.total;
    return a.dang < b.dang;
}

bool operator > (Distance a, Distance b){
    if(a.dang == b.dang) return a.total > b.total;
    return a.dang > b.dang;
}

Distance operator + (Distance a, Distance b){
    return { a.dang + b.dang, a.total + b.total };
}

bool operator > (NodeAndDist a, NodeAndDist b){
    return a.dist > b.dist;
}

int main(){
    int N, M;
    scanf("%d %d", &N, &M);

    for(int i = 0; i < M; i++){
        int u, v, d;
        scanf("%d %d %d", &u, &v, &d);
        adj[u].push_back({v, {d, 1}});
        adj[v].push_back({u, {d, 1}});
    }

    priority_queue<NodeAndDist, vector<NodeAndDist>, greater<NodeAndDist>> heap;
    for(int i = 2; i <= N; i++)
        dist[i] = {INF, INF};
    dist[1] = {0, 0};

    heap.push({1, dist[1]});
    while(!heap.empty()){
        NodeAndDist lowest = heap.top();
        heap.pop();

        int current = lowest.node;

        for(NodeAndDist to : adj[current]){
            int neighbour = to.node;
            Distance weight = to.dist;

            if(dist[current] + weight < dist[neighbour]){
                dist[neighbour] = dist[current] + weight;
                heap.push({neighbour, dist[neighbour]});
            }
        }
    }

    if(dist[N].total == INF)
        printf("-1\n");
    else
        printf("%d %d\n", dist[N].dang, dist[N].total);
}
